src/Lorentz/Contracts/BaseDAO/Types.hs

```diff
@@ -3,89 +3,351 @@
 {-# OPTIONS_GHC -Wno-orphans #-}
 
 module Lorentz.Contracts.BaseDAO.Types
-  ( Operators
+  ( Config(..)
+  , defaultConfig
+
+  , Operators
   , Ledger
+  , LedgerValue
+
   , Parameter (..)
+  , ProposeParams(..)
+  , ProposalId
+  , Proposal (..)
+  , VotingPeriod
+  , VoteParam (..)
+  , VoteType
+  , QuorumThreshold
+
   , Storage (..)
   , StorageC
   , TransferOwnershipParam
 
   , emptyStorage
   , mkStorage
+
+  , unfrozenTokenId
+  , frozenTokenId
   ) where
 
-import qualified Data.Map.Internal as Map
-import Text.Show (Show(..))
+import Universum hiding ((>>), drop)
 
+import qualified Data.Map.Internal as Map
 import Lorentz
 import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
 import Michelson.Runtime.GState (genesisAddress)
+import qualified Data.Kind as Kind
+import Util.Markdown
+
+------------------------------------------------------------------------
+-- Configuration
+------------------------------------------------------------------------
+
+data Config proposalMetadata = Config
+  { cDaoName :: Text
+  , cDaoDescription :: Markdown
+  , cUnfrozenTokenMetadata :: FA2.TokenMetadata
+  , cFrozenTokenMetadata :: FA2.TokenMetadata
+
+  , cProposalCheck :: forall s
+      . (ProposeParams proposalMetadata) & s
+    :-> Bool & s
+  -- ^ A lambda used to verify whether a proposal can be submitted.
+  -- It checks 2 things: the proposal itself and the amount of tokens frozen upon submission.
+  -- It allows the DAO to reject a proposal by arbitrary logic and captures bond requirements
+
+  , cRejectedProposalReturnValue :: forall s
+      . (Proposal proposalMetadata) & s
+    :-> ("slash_amount" :! Natural) & s
+  -- ^ When a proposal is rejected, the value that voters get back can be slashed.
+  -- This lambda specifies how many tokens will be received.
+  -- For example, if Alice freezes 100 tokens to vote and this lambda divides the value by 2,
+  -- only 50 tokens will be unfrozen and other 50 tokens will be burnt.
+
+  , cDecisionLambda :: forall s
+      . (Proposal proposalMetadata) & (Storage proposalMetadata) & s
```

Comment:
`:` should be better than `&` for the reason described in https://issues.serokell.io/issue/TM-307. Moreover, according to that issue `&` may end up being deprecated. So I think it's better to use `:` in new code.

Reply:
@gromakovsky I've updated this, add flush tests, as well as handle all the left TODOs.

2 issues that I found which I marked as `[issue 1]` and `[issue 2]` in the source code.

- [issue 1]: Since it is currently not possible to do a `delay 3 seconds` in Nettest, we cannot test if a proposal voting period is over. I have a workaround for the tests by having a `cForceVotingPeriod` in the `Config` which is just to ensured that the tests work.
  - This issue is linked with issue in morley: https://gitlab.com/morley-framework/morley/-/issues/351
  - We will remove the `cForceVotingPeriod` after the above issue is resolved. I propose we handle this in a separate issue.

- [issue 2]: It is currently not possible to check just field value in `Storage` via Nettest. This is related to `Storage` containing `BigMap`. This is the related issue in morley: https://gitlab.com/morley-framework/morley/-/issues/415#note_435327096
  - I propose we handle this in a separate issue as well.

Some other problems that I have is that, some implementations are depended on `burn` entrypoints, so I add a `TODO: [#15]` which should be handled in #15 itself. (Example: slash value need to be burn`.

The last thing is testing `cDecisionLambda`, I'm not really sure how to test this.

For the rests, everything is ready for review.

Hope to get your advice on this, Ivan

Reply:
> [issue 1]: 

I wrote [a comment](https://serokell.slack.com/archives/CF2UVU8RX/p1603736318499400?thread_ts=1603706879.484000&cid=CF2UVU8RX) about that. I think it can be implemented quite easily for Emulator, so I would wait for that.

> [issue 2]:

AFAIu a more relevant upstream issue is https://gitlab.com/morley-framework/morley/-/issues/346. Anyway, I don't mind handling it in a separate issue since the upstream issue is not started yet.

> slash value need to be `burn`

Hm, AFAIU, `burn` is basically `debitFrom`, can't you use `debitFrom`?

> The last thing is testing cDecisionLambda, I'm not really sure how to test this.

Well, you can define a decision lambda that does _something_, for example it can transfer 100 mutez to a predefined address when proposal passes. And then check that 100 mutez is actually transferred.

Reply:
I've created 2 separate issues to track the above problem.

I also add `burnFrozenToken` to make the slash function works properly now. 
I added a test for `cDecisionLambda` as well.

Related to using `ByteString` as key, I've refactored that as well. You can check it, Ivan.